### PR TITLE
gui: Option To Add a Background Image

### PR DESCRIPTION
**Pull request by 1whatleytay**
As per request, config.yml accepts the `background-image` and `background-alpha` options.

`background-image` is an optional path to some picture on your hardrive. It changes your launch page's background for some nice tasty customization. It should stretch to fit your screen, just pretend that it is in the next screenshot.
Screen Shot 2019-04-21 at 12 05 34 PM

`background-alpha` is an optional decimal that lets you see the image more or less. The higher the value, the harder it is to see. Kinda confusing imo but at least it works.

### CODE REVIEW

File: src/emulator/gui/include/gui/state.h [hunk 1]
```diff
@@ -90,6 +90,8 @@ struct State {
     char disassembly_count[5] = "100";
     std::vector<std::string> disassembly;
 
+    uint32_t background_texture = 0;
```
Comment (1whatleytay) [inline src/emulator/gui/include/gui/state.h]: Just wanted to point out that this isn't a special texture object or even a `GLuint`.

File: src/emulator/gui/src/gui.cpp [hunk 2]
```diff
@@ -120,12 +123,40 @@ static void init_font(State &gui) {
...
 
+static void init_background(State &gui, const std::string &image_path) {
+    if (!std::ifstream(image_path).good()) {
+        LOG_INFO("Invalid background file path {}.", image_path);
+        return;
+    }
+
+    int32_t width = 0;
+    int32_t height = 0;
+    stbi_uc *data = stbi_load(image_path.c_str(), &width, &height, nullptr, 3);
```
Comment (pent0) [inline src/emulator/gui/src/gui.cpp]: STBI_rgb_alpha. Waste some more GPU memory to acheive godly effect on some image

File: src/emulator/gui/src/gui.cpp [hunk 3]
```diff
@@ -120,12 +123,40 @@ static void init_font(State &gui) {
...
+
+    if (!data) {
+        LOG_INFO("Could not load background from {}.", image_path);
+        return;
+    }
+
+    glGenTextures(1, &gui.background_texture);
+    glBindTexture(GL_TEXTURE_2D, gui.background_texture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, width, height, 0, GL_RGB, GL_UNSIGNED_BYTE, data);
```
Comment (pent0) [inline src/emulator/gui/src/gui.cpp]: GL_RGBA too!

**Comment by Zangetsu38** [comment]
No relate it, but if you can found how implement game icon ^^

**Comment by 1whatleytay** [comment]
Yeah, now that we have stb included might as well do other things with it. Will look into it.

**Comment by 1whatleytay** [comment]
If you don't think we need a glDeleteTextures call that's cool. I just wanted to play it safe.

**Comment by pent0** [comment]
Everything in Vita3K should properly cleanup itself with RAII. Wraps it in a [Resource<GLint>](https://github.com/Vita3K/Vita3K/blob/master/src/emulator/util/include/util/resource.h)

**Comment by VelocityRa** [comment]
@1whatleytay Yup you should do that. More specifically use [GLObject](https://github.com/Vita3K/Vita3K/blob/master/src/emulator/glutil/include/glutil/object.h).
Will need some template magic to deal with the fact that `glDeleteTextures` accepts 2 arguments (first one being '1') and it it currently [doesn't do](https://github.com/Vita3K/Vita3K/blob/master/src/emulator/glutil/src/object.cpp#L24) that, or if you're not very experienced with them maybe you can figure out a simpler way to do it, like a deriving class where you override the deleter.

**Comment by VelocityRa** [comment]
Or maybe we should just have a separate class for it like so (rough idea):
```cpp
class GLTexture
{
public:
    GLTexture(const std::string& name, int width, int height) :
        m_id(LoadTexture(name.c_str(), width, height)) {
    }

    ~GLTexture() {
        if(m_id)
            glDeleteTextures(1, &m_id);
        m_id = 0;
    }

    GLuint id() const {
        return m_id;
    }

private:
    GLTexture(const GLTexture&);
    GLTexture& operator=(const GLTexture&);

    GLuint m_id;
};
```
in which you'd also put your `init_background` configuration GL calls so it's more reusable.

**Comment by VelocityRa** [comment]
...or here's [RPCS3's texture class](https://github.com/RPCS3/rpcs3/blob/12dc3c18723788024fd62a8dd6dd0f24d830d479/rpcs3/Emu/RSX/GL/GLHelpers.h#L1332-L1820). Maybe we ought to simply steal theirs? :)

**Comment by pent0** [comment]
Or just do [this](https://godbolt.org/z/5F-Fpk)?

**Comment by VelocityRa** [comment]
@pent0 This two-argument delete call happens often for GL objects, so maybe it's better to have a way to do it in `GLObject` itself.
Also consider `GLTexture` which would abstract the calls needed to configure the texture as I said (you _could_ make an `init` function or something if you don't like OOP, but we already have classes for GL stuff so why not).

**Comment by 1whatleytay** [comment]
What's wrong with
```
    GLuint texture;
    glGenTextures(1, &texture);
    gui.background_texture.init(texture, [](GLuint tex) { glDeleteTextures(1, &tex); });
```

**Comment by 1whatleytay** [comment]
Ehh!? My bad, I'm a bit tired today. Is this okay?

**Comment by VelocityRa** [comment]
@1whatleytay I don't think so, please read above.

**Comment by 1whatleytay** [comment]
Okay, how about this? (see amend)
Would rather use the existing GLObject class than make another one specifically for textures. I'm sure it will be helpful, but I'm not sure if this is the right PR to put it under imo.

**Comment by VelocityRa** [comment]
Ok, we can refactor later if need be.
Could you change ShaderDeleter/Deleter names to something like SingularDeleter/AggregateDeleter though?

**Comment by 1whatleytay** [comment]
No problem. Amended again.

File: src/emulator/glutil/include/glutil/object.h [hunk 4]
```diff
@@ -24,20 +24,25 @@
...
     operator GLuint() const;
+    operator bool() const;
 
 private:
+    bool init(GLuint name);
     const GLObject &operator=(const GLObject &) = delete;
 
     GLuint name = 0;
-    Deleter deleter = nullptr;
+    AggregateDeleter deleter = nullptr;
```
Comment (VelocityRa) [inline src/emulator/glutil/include/glutil/object.h]:
```suggestion
    AggregateDeleter aggregate_deleter = nullptr;
```

File: src/emulator/glutil/include/glutil/object.h [hunk 5]
```diff
@@ -24,20 +24,25 @@
...
+    SingularDeleter shaderDeleter = nullptr;
```
Comment (VelocityRa) [inline src/emulator/glutil/include/glutil/object.h]:
```suggestion
    SingularDeleter singular_deleter = nullptr;
```

File: src/emulator/glutil/include/glutil/object.h (hunk 4 above)
Comment (VelocityRa) [reply src/emulator/glutil/include/glutil/object.h]: This was not renamed

Comment (1whatleytay) [reply src/emulator/glutil/include/glutil/object.h]: AHHHHH I'm sorry T_T

**Comment by petmac** [comment]
Apologies for missing the review boat on this one. Looks like the changes to `GLObject` to add aggregate deleters are more or less what `GLObjectArray` is for. It's OK to have an array with 1 thing in it.

**Comment by VelocityRa** [comment]
Damn. Forgot about `GLObjectArray`.

**Comment by VelocityRa** [comment]
Maybe we leave the changes as they are and refactor existing code to use `GLObject` when only 1 object is required?
Some current usages of `GLObjectArray` one are for 1 object, and seeing, ie. `context.element_buffer[0]` all over in the code is misleading that more than one such objects exist (I remember being briefly confused about this in the past). We could enforce this by disallowing `GLObjectArray` to be used with 1 object.

**Pull request merged**